core/src/main/java/org/apache/brooklyn/util/core/task/ValueResolver.java

```diff
@@ -434,7 +434,11 @@ protected boolean isEvaluatingImmediately() {
                         // (should discourage this in favour of task factories which can be transiently interrupted?)
                         BrooklynTaskTags.addTagDynamically(task, BrooklynTaskTags.NON_TRANSIENT_TASK_TAG);
                     }
-                    exec.submit(task);
+                    if (timer!=null || Thread.currentThread().isInterrupted()) {
+                        exec.submit(task);
+                    } else {
+                        exec.get(task);
```

Comment:
I'm not sure how this relates to `isEvaluatingImmediately()`. For the `instanceof Future` case, that checks explicitly for immediately. Here, it will just call `exec.get()`. Previously, it would have done `submit` and then gone into the next if block because Task extends Future. I think you should guard this to only do `exec.submit` if also not `!isEvaluatingImmediately()`.

I'm guessing you might be relying on immediately having set the thread to interrupted first? I'm uncomfortable about that because we call `Tasks.resolving()...immediately(true)...getMaybe()`in a few places. That won't have gone through the `executionContext.getImmediately` code.

It's really hard to reason about whether this code will work, and how it's used, because it's all so weakly typed with `Object` and `instanceof` all over the place. If I recall correctly, `immediately` with a `Task` never quite worked properly but I can't find any appropriate tests in `ValueResolverTest`.

My general feeling is that we try to (accidentally?!) support way too many things by taking `Object` and then adding a bunch of instanceof checks, and a bunch of flags to customize the behaviour (which grow over time).

Reply:
good spot on the `immediate(true)` entry point.  i think how we resolve that is the same as the question over `getImmediately(unsubmittedTask)` semantics.  will come back to this (as part of this review but not the initial commit).

as for how we got here, immediate evaluation was retro-fitted on to many different code paths  -- some of which took `Supplier`, some which took `Task`, others `Future`, `Callable` or even `Closure`.  that's why the checks are here.  rather than immediate fail or do the wrong thing when given unexpected input, we've built up logic (not accidentally!) to try to do the right thing, and (also to the good) it is centralised here rather than all over.  i think on balance it is worth it to be able to do config validation early.  it would be good and feasible to cut down the types we use for "not-yet-known" values but the more important improvement i think was centralising it here.  venting is understandable, i want to do it myself, but of course the criticism would be more useful on the PR where the complexity was introduced rather than where I'm attempting to clean it up.  :)